src/Core/Checkout/Promotion/Cart/PromotionCalculator.php

```diff
@@ -269,12 +264,17 @@ private function calculateDiscount(LineItem $item, Cart $calculatedCart, SalesCh
         // remember our initial package count
         $originalPackageCount = $packages->count();
 
-        foreach ($calculatedCart->getLineItems() as $item) {
-            $item->setStackable(true);
-            $this->splitted[$item->getId()] = $this->lineItemQuantitySplitter->split($item, 1, $context);
+        $promotion = $this->getPromotionEntity($item->getPayloadValue('promotionId'), $calculatedCart);
+        $discountEntity = $promotion?->getDiscounts()?->get($item->getId());
```

Comment:
You can't expect to always have a promotion entity, for example, while order recalculation / pinned promotion you don't.
```suggestion
        $promotion = $this->getPromotionEntity($item->getPayloadValue('promotionId'), $calculatedCart);
        $discountEntity = $promotion?->getDiscounts()?->get($item->getId());
```
Can't you get this information from the payload of the line item discount entity?

If not, $shouldSplit should fallback to true instead of false if no promotion entity is found :thinking:

Reply:
Logic is now aligned with `Shopware\Core\Checkout\Promotion\Cart\Discount\ScopePackager\CartScopeDiscountPackager`.